### PR TITLE
Update CDVLocation.m

### PR DESCRIPTION
**Pull request by avishekcode**
Just added background mode to this function:
- (void)startLocation:(BOOL)enableHighAccuracy
{
    [self.commandDelegate runInBackground:^{

This stops iOS from giving warnings that the Plugin should be executing in background mode, and also stops random crashes of the plugin and makes it more stable on iOS.

### Platforms affected
iOS

### What does this PR do?
Increase Stability by running Function in Background Mode.

### What testing has been done on this change?
iOS tests shows that without this PR, warning is showing and also plugin crashes at some points.
With the PR plugin gains stability, and no warning found.

### CODE REVIEW

**Comment by cordova-qa** [comment]
Cordova CI Build has completed successfully.

**Commit**     - [Link](https://github.com/apache/cordova-plugin-geolocation/pull/87/commits/e56ab3bf8aab2cd5fa5942a4a28be4dc53024c83)
**Dashboard** - [Link](http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33/)

| Builder Name  | Console Output | Test Report | Device Logs  |
|     :---:     |     :---:      |   :---:     |     :---:    |
| [Windows 10  Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=windows-10-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=windows-10-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=windows-10-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=windows-10-store/artifact/) |
| [iOS 9.3]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=ios-9.3/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=ios-9.3/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=ios-9.3/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=ios-9.3/artifact/) |
| [iOS 10.0]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=ios-10.0/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=ios-10.0/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=ios-10.0/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=ios-10.0/artifact/) |
| [Android 4.4]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=android-4.4/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=android-4.4/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=android-4.4/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=android-4.4/artifact/) |
| [Android 5.1]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=android-5.1/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=android-5.1/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=android-5.1/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-geolocation-pr/33//PLATFORM=android-5.1/artifact/) |

**Comment by filmaj** [comment]
Ping @shazron for a review.

@avishekcode have you signed an Apache ICLA?

**Comment by avishekcode** [comment]
@filmaj Nope. Is that a requirement? I didn't think there would be any licensing issue, since I am contributing to an already open project, where the terms are already in place. I would think licensing might be required if I created a entirely new project.

**Comment by filmaj** [comment]
@avishekcode in general it is required, I believe there is some wiggle room for trivial contributions, but in this case, I think the changes are significant enough that we'll need an ICLA signed from you.

Check out: https://www.apache.org/licenses/#clas

If you wouldn't mind filling out + signing the ICLA here: https://www.apache.org/licenses/icla.pdf and e-mailing it to secretary@apache.org, then I think that's all we need to safely pull in your code.

Thanks so much!

**Comment by avishekcode** [comment]
@filmaj Greatly appreciate for letting me know.

I just submitted the Individual Contributor License Agreement to secretary@apache.org via email.

Not sure if I need to do anything else.

Thanks!

**Comment by filmaj** [comment]
Just received the ICLA. Thanks!

**Comment by filmaj** [comment]
Hey @avishekcode, I'm trying to reproduce these warnings you mentioned:

> This stops iOS from giving warnings that the Plugin should be executing in background mode, and also stops random crashes of the plugin and makes it more stable on iOS.

How do they look like? I cannot seem to reproduce with the latest master release of this plugin (without your code changes). I am simply issuing a `getCurrentLocation` call when my app starts up.

File: src/ios/CDVLocation.m [hunk 1]
```diff
@@ -94,61 +94,62 @@ - (BOOL)isLocationServicesEnabled
...
-        BOOL authStatusAvailable = [CLLocationManager respondsToSelector:@selector(authorizationStatus)]; // iOS 4.2+
-        if (authStatusAvailable) {
-            NSUInteger code = [CLLocationManager authorizationStatus];
-            if (code == kCLAuthorizationStatusNotDetermined) {
-                // could return POSITION_UNAVAILABLE but need to coordinate with other platforms
-                message = @"User undecided on application's use of location services.";
-            } else if (code == kCLAuthorizationStatusRestricted) {
-                message = @"Application's use of location services is restricted.";
+    [self.commandDelegate runInBackground:^{
+        if (![self isLocationServicesEnabled]) {
```
Comment (shazron) [inline src/ios/CDVLocation.m]: You have to be careful here -- you can't use self in the block, you have to grab a weak reference to self and use it in the block. Like so: https://github.com/apache/cordova-ios/blob/92a62997adb3c8512328d5a0ae42fe5d156291f1/CordovaLib/Classes/Private/Plugins/CDVHandleOpenURL/CDVHandleOpenURL.m#L58

Just us a general note -- I don't notice any of this in your code but it may occur in nested calls -- any calls to UIKit framework classes must occur in the main thread while called inside the block, so that may complicate things.

**Comment by shazron** [comment]
@avishekcode what device are you testing it on? @filmaj I believe if he is using an older phone, it might affect timings

**Comment by filmaj** [comment]
@shazron I've been trying to reproduce these warnings on a pretty old iOS device: iPhone 5S, running iOS 10.2. Haven't been able to repro anything that looks weird in the device console.

**Comment by shazron** [comment]
Hmm not sure then :/ 

Regarding my line comments on this PR, I realize that we have another instance not following the rules in the existing code 😱  https://github.com/apache/cordova-plugin-geolocation/blob/9f9496f0577ff66fb2449ac21b28ee1f0252aca2/src/ios/CDVLocation.m#L197

**Comment by avishekcode** [comment]
@shazron @filmaj 
Hi!

I literally, just downloaded the current plugin, and without my modification, I get these warnings constantly via Xcode, while testing on iPhone 5s:
2017-04-26 17:45:32.151697-0400 TabCab Inc[3927:2025999] THREAD WARNING: ['Geolocation'] took '10.250000' ms. Plugin should use a background thread.
2017-04-26 17:45:32.256468-0400 TabCab Inc[3927:2025999] THREAD WARNING: ['Geolocation'] took '33.067871' ms. Plugin should use a background thread.
2017-04-26 17:45:52.700650-0400 TabCab Inc[3927:2025999] THREAD WARNING: ['Geolocation'] took '25.637939' ms. Plugin should use a background thread.
2017-04-26 17:46:18.191620-0400 TabCab Inc[3927:2025999] THREAD WARNING: ['Geolocation'] took '10.547852' ms. Plugin should use a background thread.

Device: iPhone 5s, but I believe it happens even in newer phones, if I am using the plugin for applications like Uber (Motion capture in the background).

Now, I apologize if I broke some rules while trying to implement the solution. I will try to dig into the rules, and any quick suggestion how to fix the problem otherwise?

My solution puts the function in background thread, stops all warnings, and it increases the performance of the plugin and the overall app.

My use of the plugin in my applications, is as follows:
function getAccurateGPS(GPSok,GPSerr,GPSing,options,Abort)
{
	var lastPos,locECnt=0,watchID,navTim;options=options || {};
	if(Abort){clearTimeout(navTim);navTim=null;navigator.geolocation.clearWatch(watchID);}
	else
	{	var checkLocation=function(position)
		{	lastPos=position;locECnt=locECnt+1;	//Ignore first event unless it's the only one because some devices send cached location even with maxaimumAge=0
			if((position.coords.accuracy<=options.desiredAccuracy) && (locECnt>1)){clearTimeout(navTim);navTim=null;navigator.geolocation.clearWatch(watchID);foundPos(position);}else{GPSing(position);}
		};
		var stopTrying=function(){navigator.geolocation.clearWatch(watchID);foundPos(lastPos);};
		var onError=function(error){clearTimeout(navTim);navTim=null;navigator.geolocation.clearWatch(watchID);GPSerr(error);};
		var foundPos=function(position){GPSok(position);};
		if(!options.maxWait)			options.maxWait=10000;
		if(!options.desiredAccuracy)	options.desiredAccuracy=5;
		if(!options.timeout)			options.timeout=options.maxWait;
		options.maximumAge=0;options.enableHighAccuracy=true;
		watchID=navigator.geolocation.watchPosition(checkLocation,onError,options);
		navTim=setTimeout(stopTrying,options.maxWait);//Timeout
	}
}

Basically, my applications require constant motion tracking, and I try to capture the most accurate position available.

Please let me know if there is anything else you need from me, or if you have any suggestions.

Thanks!

**Comment by filmaj** [comment]
Okay, with that sample code of using the watch, I am able to reproduce one warning of this on an iPhone 5S running iOS 10.2:

```
2017-04-26 17:14:37.015628 HelloCordova[273:48525] THREAD WARNING: ['Geolocation'] took '44.689941' ms. Plugin should use a background thread.
```

I need to run the app from xcode to see that warning, though - that warning does not show in the Device Console.

Now when using @avishekcode's patch, the warning is gone.

So I'm fine with merging this patch in as long as any comments from @shazron are addressed.

File: src/ios/CDVLocation.m (hunk 1 above)
Comment (jcesarmobile) [inline src/ios/CDVLocation.m]: I think we are abusing of the weak references to self in the plugins. In most cases it's not really needed because we don't retain the block http://stackoverflow.com/questions/20030873/always-pass-weak-reference-of-self-into-block-in-arc

Comment (shazron) [inline src/ios/CDVLocation.m]: It's a good practice (yes I read the article). In this case, self will always be retained (it's the plugin, which is always retained until the app closes).
